### ISSUE TITLE
Improve Local Group satellite incompleteness model

### ISSUE THREAD

**Issue by abensonca** [galacticusorg/galacticus#73]
Use the algorithm from [Drlica-Wagner et al. (2019)](https://arxiv.org/abs/1912.03302).

**Comment by abensonca** [comment]
See [Tan et al. (2025)](https://arxiv.org/abs/2509.12313).

**Comment by abensonca** [comment]
## Assessment

**Verdict: worth doing, and tractable — but implement against [Tan et al. (2025)](https://arxiv.org/abs/2509.12313) (DELVE MW Census I) rather than [Drlica-Wagner et al. (2020)](https://arxiv.org/abs/1912.03302).** The DELVE census supersedes DES Y3+PS1 (DES Y6 + DELVE DR3 + PS1 DR1; ~91% of the |b| ≥ 15° sky), uses the same interface for its selection function, and ships a cleaner public data release. The DW+20 products remain available in the identical format, so they are a useful cross-check rather than the primary target.

### What we do now

Three survey geometries are combined in `outputAnalysisLocalGroupStellarMassFunction` (`source/output/analyses/Local_Group/stellar_mass_function.F90:348`):

| class | detectability model |
|---|---|
| `surveyGeometryLocalGroupClassical` | detected iff $M_\star > 10^5\,M_\odot$, all-sky minus a Zone-of-Avoidance mangle mask |
| `surveyGeometryLocalGroupSDSS` | $d_\mathrm{max} = 15.7\,\mathrm{kpc}\,(M_\star/100\,M_\odot)^{0.51}$ (Kim, Peter & Hargis 2018) |
| `surveyGeometryLocalGroupDES` | $d_\mathrm{max} = 11.3\,\mathrm{kpc}\,(M_\star/100\,M_\odot)^{0.48}$ (polynomial fit to Drlica-Wagner et al. 2015 efficiencies) |

Four limitations, in decreasing order of importance:

1. **No size dependence.** Detectability of ultra-faints is a strong function of surface brightness, i.e. of $r_{1/2}$ at fixed $M_V$; at large $r_{1/2}$ a satellite is undetectable at *any* distance. We currently ignore this entirely. This matters doubly because we already constrain the model's mass–size relation (`localGroupMassSizeRelation`), so the same predicted sizes should be feeding the completeness correction.
2. **Binary, not probabilistic.** The real selection function is a smooth $P_\mathrm{det} \in [0,1]$; a hard $d_\mathrm{max}$ cut both biases and adds noise near the threshold.
3. **Luminosity is faked.** Both $d_\mathrm{max}$ relations assume $M_\star/L_V = 1$; the model can compute $M_V$ properly.
4. **Sample is small and the footprints are dated.** The comparison sample is built by selecting `discoverySurvey ∈ {classical, SDSS, DES}` from the Local Group database — 39 galaxy-classified systems within 300 kpc. The DELVE census sample is **51** systems within 300 kpc, with a rigorously defined, published selection function attached.

### What the published selection functions actually provide

I checked the release at https://github.com/delve-survey/delve_mw_census (Zenodo DOI 10.5281/zenodo.18383157). It is much more usable than I expected:

* Per-survey XGBoost classifiers (`classifier_stellar_{desy6,delvedr3,ps1}*.model`) taking **only four features**: $\log_{10}(D_\mathrm{helio}/\mathrm{kpc})$, $M_V$, $\log_{10}(r_{1/2}/\mathrm{kpc})$ (physical, azimuthally averaged half-light radius), and local stellar density $\rho_\star$ (stars arcmin$^{-2}$, $r<22$).
* HEALPix footprint/bad-region masks and $\rho_\star$ maps (nside 128) per survey.
* Combination rule from their Notebook 1: $P_\mathrm{det} = P_\mathrm{DES} + P_\mathrm{DELVE} + P_\mathrm{PS1}$ (footprints disjoint after masking), with $M_V < -12.5 \Rightarrow P = 1$ and $D_\mathrm{helio} < 16$ kpc or $> 400$ kpc $\Rightarrow P = 0$.
* `offical_delve_census_refgalaxy_list.csv` — the census sample itself, with `delve_census` flag, $M_V$, $r_{1/2}$, $D_\mathrm{helio}$, $D_\mathrm{GC}$, and `dwarf_class`. This is the target sample, ready-made.

### The key point that makes this easy

We must not put XGBoost + healpy in the Galacticus runtime. We do not have to: **our satellite positions have no meaningful orientation relative to the Galactic plane or the survey footprints anyway** (today's mangle masks are applied to raw model coordinates, which is a random-orientation assumption in disguise). So we can marginalise the sky dependence offline.

For a satellite at galactocentric radius $r$, assume its direction $\hat{n}$ is isotropic. That direction fixes both the heliocentric distance ($D^2 = r^2 + R_0^2 - 2 r R_0 \cos\theta$, $R_0 = 8.122$ kpc) *and* the sky position, hence the mask value and $\rho_\star$. Averaging over $\hat{n}$ gives

$$\langle P_\mathrm{det}\rangle(r, M_V, r_{1/2}) = \frac{1}{4\pi}\int P_\mathrm{det}\big(D(\hat{n}), M_V, r_{1/2}, \rho_\star(\hat{n})\big)\,\big[1-\mathrm{mask}(\hat{n})\big]\,\mathrm{d}\Omega ,$$

a **3-D table** (~50×50×50 ≈ 1 MB) that we precompute once and interpolate trilinearly at runtime. This simultaneously:

* removes any Python/ML dependency from the executable;
* handles the Sun's offset correctly (we currently use galactocentric distance where the selection function wants heliocentric);
* folds in the Zone of Avoidance and the survey footprints exactly, replacing the mangle masks for this analysis;
* eliminates the Monte-Carlo noise from testing individual satellites against a fixed footprint.

## Implementation plan

**1. Offline table builder** — `scripts/aux/localGroupSelectionFunction.py` (or in `galacticusorg/datasets`).
Loads the three classifiers, masks, and density maps; evaluates $\langle P_\mathrm{det}\rangle$ on a grid in $\log_{10} r_\mathrm{GC}$ (16–400 kpc), $M_V$ ($-18$ to $+2$), $\log_{10} r_{1/2}$ (2–3000 pc); writes HDF5 to `GALACTICUS_DATA_PATH/static/observations/localGroup/`, with provenance attributes (paper, Zenodo DOI, model file versions, $R_0$, grid definition). Requires `xgboost`/`healpy` only at build time. Also emit a DW+20 variant from `des-science/mw-sats` for cross-checking.

**2. New Fortran class for the tabulated selection function.**
A `satelliteDetectionProbability` functionClass (or, more cheaply, a single concrete class) with a `tabulated` implementation reading the HDF5 file and doing trilinear interpolation via three `interpolator` objects using `linearFactors`/`linearWeights` (`source/numerical/interpolation/1D.F90:238`). Extrapolation clamped to $P = 0$ outside the table.

**3. Hook it into the analyses as a weight, not a filter.**
Add `outputAnalysisWeightOperatorLocalGroupDetection`, which extracts from the node:
* $M_V$ — via `nodePropertyExtractorLuminosityStellar`/`absolute_magnitude` in `Buser_V` (AB→Vega offset for $V$ is ~0.02 mag; apply it, but it is negligible);
* $r_{1/2}$ — `nodePropertyExtractorRadiusHalfMassStellar` converted 3-D → projected ($\approx 0.75\times$; expose as a parameter);
* $r_\mathrm{GC}$ — magnitude of the orbital position;

and returns `weight × P_det`. The weight route slots straight into `volumeFunction1DAnalyze` (`source/output/analyses/volume_function_1d.F90:728`) and is lower variance than a stochastic filter. Worth noting: the covariance accumulated there scales as $w^2$ per object, which omits the $P(1-P)$ Bernoulli term — but `localGroupStellarMassFunctionLogLikelihood` does not use the covariance (it uses the negative-binomial with `negativeBinomialScatterFractional`), so this is cosmetic and can be addressed later.

Replace `galacticFilterSurveyGeometry` + `surveyGeometryCombined` in the satellite filter chain with a plain radial cut (16–300 kpc, to match the census validity range).

**4. Target sample.**
Add a census-membership flag to the Local Group database (e.g. `<census value="DELVE-MW-I"/>`), populated from `offical_delve_census_refgalaxy_list.csv` in `localGroupSatellitesUpdate.py`, and select on it in the analysis instead of on `discoverySurvey`. Selecting by `discoverySurvey` is fragile anyway — the DB currently has `DELVE`, `DELVE DR3`, `DECam Local Volume Exploration Survey`, `SDSS`, `SDSS DR8`, `SDSS DR10` as distinct strings.

**5. Add a luminosity-function analysis binned in $M_V$.**
The census is defined in $M_V$, and for ultra-faints the database's $M_\star$ is itself derived from $M_V$ with an assumed $M/L$. A `localGroupLuminosityFunction` analysis in $M_V$ is the more direct comparison; keep the stellar mass function alongside it.

**6. Model configuration and validation.**
Add `<luminosityFilter value="Buser_V"/>` (rest-frame, $z=0$) to `testSuite/parameters/validate_milkyWay.xml` and `parameters/baryonicPhysicsConstrained.xml` — one extra filter, cheap. Regenerate validation baselines and update `scripts/build/ghPagesAnalysisThresholds.yml` (`localGroupStellarMassFunction` and any new entry).

Keep `surveyGeometryLocalGroup{Classical,SDSS,DES}` in place so the old analysis remains reproducible.

Rough sizing: ~1 week total — the table builder and the Fortran interpolation class are each about a day; the analysis/DB/parameter work and revalidation are the bulk.

## Risks and decisions needed

* **Licensing.** `delve_mw_census` carries a Zenodo DOI and `citation.cff` but no `LICENSE` file (`des-science/mw-sats` is GPL-3.0). Before shipping a derived table in `galacticusorg/datasets` we should confirm redistribution is fine — probably a one-line email.
* **The constraint now couples to the predicted sizes.** This is physically correct, but it means an error in the model's dwarf sizes propagates directly into the predicted luminosity function. Before adopting this as a constraint we should check the model's $r_{1/2}$ distribution for ultra-faints against `localGroupMassSizeRelation`.
* **Classification.** The census sample is `dwarf_class ∈ {D, PD}` (45 confirmed + 7 probable dwarfs); we should decide whether to include the probables, and make sure it is consistent with what the model counts as a galaxy.
* **Anisotropy.** Tan et al. report a ~2σ anisotropy associated with LMC-clustered satellites. The sky-marginalised treatment above explicitly assumes isotropy, so this is out of scope — worth stating in the class documentation.
* **Cheaper fallback**, if the above is too much at once: keep the hard-cutoff geometries but refit $d_\mathrm{max}$ as a function of *both* $M_V$ and $r_{1/2}$ from the 50% contour of the classifier. That captures limitation 1 (the important one) for much less work, at the cost of leaving 2–4 unaddressed.

---
*Assessment prepared with assistance from Claude Code; the reference material and code paths above were checked against the current tree, but the plan has not been implemented or tested.*

**Comment by abensonca** [comment]
## Open decisions — resolved

Recording the resolutions to the four items raised above, so the plan is unambiguous when it comes to be implemented.

1. **Licensing — not a blocker.** The products derive from a published paper and re-use is fine provided credit is given. Step 1 therefore ships the derived table into `galacticusorg/datasets` directly; the provenance attributes on the HDF5 file (paper reference, arXiv ID, Zenodo DOI `10.5281/zenodo.18383157`, classifier/mask file versions) become the attribution mechanism, and the new class documentation should carry a `:cite:p:` reference to Tan et al. (2025).

2. **Predicted dwarf sizes — good enough to use.** Our sizes are at least reasonably correct given the observational uncertainties, so no pre-check gate is needed before adopting the size-dependent weighting. Worth still reporting the model $r_{1/2}$ distribution for ultra-faints alongside the first validation run, simply because the selection function is steep in $r_{1/2}$ and it will be the first thing to inspect if the luminosity function moves unexpectedly.

3. **Census classification — include probable dwarfs.** So the selection is `delve_census == True` with `dwarf_class ∈ {D, PD}`: 44 confirmed + 7 probable = **51 systems within 300 kpc** (this is the figure already quoted above — restricting to confirmed dwarfs only would give 44). The census-membership flag added to the Local Group database should record the class as well as the flag, so the confirmed-only variant remains selectable without re-deriving anything.

4. **Anisotropy — proceed with isotropy, revisit later.** The sky-marginalised $\langle P_\mathrm{det}\rangle(r_\mathrm{GC}, M_V, r_{1/2})$ table stands as the implementation. Separate work on constrained Milky Way systems with an LMC in the correct location is under way; once that lands the isotropy assumption is the thing to revisit, since a satellite population with a genuine LMC association *does* have a meaningful orientation relative to the survey footprints.

   To keep that door open at essentially zero cost now: the offline builder should retain the un-marginalised evaluation path (the full $P_\mathrm{det}(D_\mathrm{helio}, M_V, r_{1/2}, \rho_\star)$ classifier call plus the HEALPix mask/density lookup), so a future direction-dependent variant is a change of output table rather than a rewrite. The Fortran side would then gain a second implementation of the same class that takes a sky direction — the weight-operator interface above does not change.
